### ISSUE TITLE
runtime/build-time dependencies

### ISSUE THREAD

**Issue by domenkozar** [cachix/cachix#9]
Currently cachix only uploads runtime closure, but one could upload all build-time dependencies using:

```
nix-store -qR --include-outputs /nix/store/blabla.drv
```

**Comment by domenkozar** [comment]
@dtzWill what about `cachix push --build-dependencies`?

**Comment by dtzWill** [comment]
(D'oh sorry for missing this issue!)

That sounds good to me! Would it be reasonable/possible to make it a config option? Not critical but while we're talking wish-list... :grin:

**Comment by domenkozar** [comment]
Config will come with #14 :)

**Comment by dtzWill** [comment]
FWIW I'm not sure including all build dependencies is quite the same, since build dependencies don't necessarily include the other outputs produced by the derivation that created the path in question--although the command in the issue description will do this (if fed a derivation, presumably obtained with `nix-store -q --deriver` as needed).

Perhaps it's just a minor terminology quibble but just wanted to make sure whatever was done made an intentional decision in this regard (my vote is forr "sibling" outputs included, but it seems this is what you were intending anyway :)).

**Comment by domenkozar** [comment]
I agree, correct term is something like `realized build closure`. I'd like to avoid these complex terms if possible, maybe add something like `--full-closure` which would have docs to describe what it means? :)

**Comment by dtzWill** [comment]
That sounds pretty much perfect to me! :)

**Comment by pjan** [comment]
Very much looking forward to this feature; rebuilding ghc for every new haskell project when using `nixpkgs-unstable` gets tiring very quickly ;-)

`--full-closure` (or even just `--full` for that matter) sounds perfect.

**Comment by domenkozar** [comment]
For security reasons I'd like to do the following (leave sources out, except for fixed output derivations):

```
nix-store -qR --include-outputs $(nix-store -q --deriver /nix/store/ilfhdrrigxdhys3lkr6bwjsmf42s0n9x-cachix-server-0.1.0.0) > with-outputs
nix-store -qR  $(nix-store -q --deriver /nix/store/ilfhdrrigxdhys3lkr6bwjsmf42s0n9x-cachix-server-0.1.0.0) > no-outputs
cat  with-outputs  no-outputs | sort | uniq -u
```

This is essentially a cached deployment without derivation files and local paths.

How does this sound?

**Comment by nomeata** [comment]
I am doing this manually with
```
nix-store -qR --include-outputs /nix/store/blabla.drv
```
now, but it is uploading stuff already cached on some other cache (`https://hydra.iohk.io` and `https://cache.nixos.org`).

Is there a way to upload only those build  dependencies not found on the other caches?

**Comment by domenkozar** [comment]
Unfortunately not yet - see https://github.com/cachix/cachix/issues/16

**Comment by domenkozar** [comment]
See also `pkgs.closureInfo`

**Comment by domenkozar** [comment]
Refs https://www.reddit.com/r/NixOS/comments/llf6fl/callcabal2nix_cachix/
